### DISCUSSION TITLE
How to use?

### DISCUSSION THREAD

**Discussion by VIE-Flamingo** [wintercms/wn-blocks-plugin#38] [Q&A] [answered]
How to render blocks in form widget? 

1. I've tried to add the these in yaml but I only see a repeater form field with a prompt, when I clicked to add item, there is nothing, just a blank menu, even when I left only type: blocks, it ran well without any exceptions, I don't know I did the correct way or not.
```yaml
 buttons:
        label: Buttons
        span: full
        type: blocks
        allow:
            - button
```

2. After I failed with the first try, I try to render it directly in PHP like this:
```php
<?php \Winter\Blocks\Classes\Block::render('title', [
    'content' => 'Lorem ipsum dolor sit amet.',
    'alignment_x' => 'left',
    'size' => 'h1',
]); ?>
```
and I got this error, it belongs to getDefaultConfig() of /plugins/winter/blocks/classes/Block.php

<img width="532" alt="Screenshot 2024-09-13 at 16 42 03" src="https://github.com/user-attachments/assets/0aae384e-d10c-4dee-93ac-1e16efc487ce">

Can some one please tell me what I should do to use Block? Or maybe a demo for how to set all things up and what the result will be if I set up like that? 

Thank you!

**Comment by bennothommo** [comment, marked as answer]
@khutachan might need a bit more info here.

- Where are you trying to use the `blocks` type? Is this in a page or a layout?
- If you change the `button` under `allow` to `pages` instead, does this work? It appears our default blocks provided by this plugin use the tag `pages`.
- Have you created your own custom blocks in your theme or any plugins?

**Reply by VIE-Flamingo** [reply]
Thank you for your response. Since I'm not entirely sure how to use this feature, I'll explain my purpose, and maybe you can point out what I'm missing.

I want to use blocks in my theme. I have several ideas on how to configure block types, but I'm unsure where to begin, so I've tried a few different approaches:

- I added the script below the `{% snowboard all %}` in the layout. I'm wondering if I can render a block without using the Winter.Page plugin? Specifically, I want to render blocks directly into the theme page, similar to how components are added to a page. Currently, I'm working without the Winter.Page plugin, but I also tried using it (I added the Twig variable as per the documentation). However, I didn’t see any block repeater form fields.

- I tried to render the block type that I registered in my plugin (I created a `.block` file, copied the title block example, and registered it in `Plugin.php`). This failed, so I reverted to using the registered `buttons` block, but that also failed.

Here’s what I did:

1. **How I registered a new block:**

   ![block registration screenshot](https://github.com/user-attachments/assets/4bffc1dc-610e-4472-bba6-42faabef3afc)

   ![block file screenshot](https://github.com/user-attachments/assets/b5571f2a-0cf1-4828-a7eb-8e9b5d4bc1d9)

   ![plugin registration screenshot](https://github.com/user-attachments/assets/9f9ec5b7-72ba-4b46-a60c-97d34c71059a)

2. **How I tried to use it in my theme:**

   ```twig
   {{ renderBlock('title', {
       'content': 'Lorem ipsum dolor sit amet.',
       'alignment_x': 'left',
       'size': 'h1',
   }) }}

3. **Error encountered**: 
An exception was thrown because `$config` was null, causing the `array_key_exists()` function to throw an exception. After investigating, I found that `$config` is null because there are no registered blocks. I checked `BlockManager::instance()->getConfigs()` and it returned an empty array.

<img width="821" alt="Screenshot 2024-09-16 at 11 13 54" src="https://github.com/user-attachments/assets/11d16b97-f5f5-4fe1-a8d3-ef85fa1c2553">

**Comment by VIE-Flamingo** [comment]
@bennothommo I think I have found the problem, everything of this plugin is running well, I can't render the block because I have booted my theme in Plugin.php `boot()` before block is booted into theme (I think so), so do you have a solution for booting an active theme within block and my code?

Here is the function that I call in `boot()`
```php 
 public function bootThemeLanguageKey()
    {
        $activeTheme = \Cms\Classes\Theme::getActiveTheme();
        $path = $activeTheme->getPath() . DIRECTORY_SEPARATOR . 'lang';
        $nameSpace = 'themes.' . $activeTheme->getId();
        $this->loadTranslationsFrom($path, $nameSpace);
    }
```

**Reply by LukeTowers** [reply]
@khutachan submit that as a separate question / issue please.
